File: Sources/ReminderCore/Music.swift
import Foundation

/// What music, if any, a reminder starts when it fires.
///
/// Three states rather than a plain on/off switch: most reminders that want
/// music want *the* playlist the user configured once in Preferences, but a
/// reminder is free to bring its own. Modelling "use the default" explicitly
/// means changing the default in Preferences updates every reminder that opted
/// in, instead of leaving stale copies of a URI scattered across reminders.
public enum MusicChoice: Codable, Equatable, Sendable {
    /// Fire silently. The default for every reminder.
    case none
    /// Play whatever playlist is set in Preferences.
    case defaultPlaylist
    /// Play this specific playlist, ignoring the Preferences default.
    case playlist(uri: String)

    /// The playlist to actually play, resolved against the global settings.
    /// `nil` means "play nothing" — either the reminder opted out, or it asked
    /// for the default and no default is configured.
    public func resolvedURI(defaultPlaylistURI: String?) -> String? {
        switch self {
        case .none:
            return nil
        case .defaultPlaylist:
            guard let uri = defaultPlaylistURI, !uri.isEmpty else { return nil }
            return uri
        case .playlist(let uri):
            return uri.isEmpty ? nil : uri
        }
    }

    /// Whether music is switched on at all, for driving a toggle in the UI.
    public var isEnabled: Bool {
        if case .none = self { return false }
        return true
    }
}

/// Parsing and validation for Spotify playlist references.
///
/// Kept in the core rather than beside the AppleScript so it can be tested
/// without AppKit, and so exactly one definition of "a URI we will run"
/// exists — the validator here is what makes interpolating a user-supplied
/// string into a script source safe.
public enum SpotifyURI {

    /// Spotify IDs are base-62 with `-` and `_` also appearing in share links.
    /// Anything outside this set is rejected, which is what keeps a quote or a
    /// script fragment from ever reaching the AppleScript source.
    private static let idPattern = "[A-Za-z0-9_-]+"

    /// The kinds of thing we are willing to play. Artists are deliberately
    /// absent: an artist URI is neither a playable track nor a valid playback
    /// context in Spotify's AppleScript dictionary, so accepting one would
    /// store a value that can only fail an hour later mid-reminder.
    private static let kinds = ["playlist", "album", "track"]

    /// Accepts a `spotify:playlist:ID` URI or an
    /// `https://open.spotify.com/playlist/ID?si=...` share link (including the
    /// localized `/intl-de/playlist/ID` form) and returns the canonical URI.
    ///
    /// Returns `nil` for anything it does not recognise, so the settings UI can
    /// reject bad input inline rather than failing an hour later mid-reminder.
    public static func normalize(_ input: String) -> String? {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        for kind in kinds {
            // Two accepted shapes, matched separately rather than with one
            // loose pattern:
            //
            //   spotify:playlist:ID              — the URI form
            //   open.spotify.com/playlist/ID     — the share link, which may
            //                                      carry a locale segment
            //                                      (/intl-de/playlist/ID) and
            //                                      a ?si=... query
            //
            // A single `spotify[:/]kind` pattern does not cover the link: there
            // the host is `spotify.com`, so `spotify` is followed by `.com/`
            // rather than a separator.
            let patterns = [
                "spotify:\(kind):(\(idPattern))",
                "spotify\\.com/(?:[A-Za-z0-9-]+/)?\(kind)/(\(idPattern))",
            ]

            for pattern in patterns {
                guard let range = trimmed.range(
                    of: pattern, options: [.regularExpression, .caseInsensitive]
                ) else { continue }

                // The ID is everything after the final separator of the match.
                let matched = String(trimmed[range])
                guard let id = matched
                    .components(separatedBy: CharacterSet(charactersIn: ":/"))
                    .last, !id.isEmpty else { continue }

                return "spotify:\(kind):\(id)"
            }
        }
        return nil
    }

    /// Whether `uri` is exactly a canonical URI we will execute.
    ///
    /// `play` validates with this immediately before building the script, so a
    /// value that reached storage by some other route (a hand-edited data file)
    /// still cannot inject AppleScript.
    public static func isValid(_ uri: String) -> Bool {
        let pattern = "^spotify:(\(kinds.joined(separator: "|"))):\(idPattern)$"
        return uri.range(of: pattern, options: .regularExpression) != nil
    }

    /// A short human label for a URI, for showing in the UI.
    public static func describe(_ uri: String) -> String {
        let parts = uri.components(separatedBy: ":")
        guard parts.count == 3 else { return uri }
        return "\(parts[1].capitalized) · \(parts[2])"
    }
}

File: Windows/Pauselet.Core/Music.cs
using System.Text.RegularExpressions;

namespace Pauselet.Core;

/// <summary>
/// What music, if any, a reminder starts when it fires.
///
/// Three states rather than a plain on/off switch: most reminders that want
/// music want *the* playlist the user configured once in Preferences, but a
/// reminder is free to bring its own. Modelling "use the default" explicitly
/// means changing the default in Preferences updates every reminder that opted
/// in, instead of leaving stale copies of a URI scattered across reminders.
/// </summary>
public abstract record MusicChoice
{
    private MusicChoice() { }

    /// <summary>Fire silently. The default for every reminder.</summary>
    public static readonly MusicChoice None = new NoneChoice();
    /// <summary>Play whatever playlist is set in Preferences.</summary>
    public static readonly MusicChoice DefaultPlaylist = new DefaultPlaylistChoice();
    /// <summary>Play this specific playlist, ignoring the Preferences default.</summary>
    public static MusicChoice Playlist(string uri) => new PlaylistChoice(uri);

    public sealed record NoneChoice : MusicChoice;
    public sealed record DefaultPlaylistChoice : MusicChoice;
    public sealed record PlaylistChoice(string Uri) : MusicChoice;

    /// <summary>
    /// The playlist to actually play, resolved against the global settings.
    /// <c>null</c> means "play nothing" — either the reminder opted out, or it
    /// asked for the default and no default is configured.
    /// </summary>
    public string? ResolvedUri(string? defaultPlaylistUri) => this switch
    {
        NoneChoice => null,
        DefaultPlaylistChoice =>
            string.IsNullOrEmpty(defaultPlaylistUri) ? null : defaultPlaylistUri,
        PlaylistChoice playlist => playlist.Uri.Length == 0 ? null : playlist.Uri,
        _ => null,
    };

    /// <summary>Whether music is switched on at all, for driving a toggle in the UI.</summary>
    public bool IsEnabled => this is not NoneChoice;
}

/// <summary>
/// Parsing and validation for Spotify playlist references.
///
/// Kept in the core so exactly one definition of "a URI we will run" exists.
/// On macOS the validator is what makes interpolating a user-supplied string
/// into AppleScript source safe; the same discipline is kept here so the data
/// files stay interchangeable and a future Windows integration inherits it.
/// </summary>
public static class SpotifyUri
{
    /// <summary>
    /// Spotify IDs are base-62 with <c>-</c> and <c>_</c> also appearing in
    /// share links. Anything outside this set is rejected, which is what keeps
    /// a quote or a script fragment from ever reaching an interpreter.
    /// </summary>
    private const string IdPattern = "[A-Za-z0-9_-]+";

    /// <summary>
    /// The kinds of thing we are willing to play. Artists are deliberately
    /// absent: an artist URI is neither a playable track nor a valid playback
    /// context, so accepting one would store a value that can only fail an hour
    /// later mid-reminder.
    /// </summary>
    private static readonly string[] Kinds = ["playlist", "album", "track"];

    /// <summary>
    /// Accepts a <c>spotify:playlist:ID</c> URI or an
    /// <c>https://open.spotify.com/playlist/ID?si=...</c> share link (including
    /// the localized <c>/intl-de/playlist/ID</c> form) and returns the
    /// canonical URI.
    ///
    /// Returns <c>null</c> for anything it does not recognise, so the settings
    /// UI can reject bad input inline rather than failing an hour later
    /// mid-reminder.
    /// </summary>
    public static string? Normalize(string input)
    {
        var trimmed = input.Trim();
        if (trimmed.Length == 0) return null;

        foreach (var kind in Kinds)
        {
            // Two accepted shapes, matched separately rather than with one
            // loose pattern:
            //
            //   spotify:playlist:ID              — the URI form
            //   open.spotify.com/playlist/ID     — the share link, which may
            //                                      carry a locale segment
            //                                      (/intl-de/playlist/ID) and
            //                                      a ?si=... query
            //
            // A single `spotify[:/]kind` pattern does not cover the link: there
            // the host is `spotify.com`, so `spotify` is followed by `.com/`
            // rather than a separator.
            string[] patterns =
            [
                $"spotify:{kind}:({IdPattern})",
                $@"spotify\.com/(?:[A-Za-z0-9-]+/)?{kind}/({IdPattern})",
            ];

            foreach (var pattern in patterns)
            {
                var match = Regex.Match(trimmed, pattern, RegexOptions.IgnoreCase);
                if (!match.Success) continue;

                // The ID is everything after the final separator of the match.
                var id = match.Value.Split(':', '/').Last();
                if (id.Length == 0) continue;

                return $"spotify:{kind}:{id}";
            }
        }
        return null;
    }

    /// <summary>
    /// Whether <paramref name="uri"/> is exactly a canonical URI we will
    /// execute. Playback validates with this immediately before acting, so a
    /// value that reached storage by some other route (a hand-edited data file)
    /// still cannot inject anything.
    /// </summary>
    public static bool IsValid(string uri)
    {
        var pattern = $"^spotify:({string.Join("|", Kinds)}):{IdPattern}$";
        return Regex.IsMatch(uri, pattern);
    }

    /// <summary>A short human label for a URI, for showing in the UI.</summary>
    public static string Describe(string uri)
    {
        var parts = uri.Split(':');
        if (parts.Length != 3) return uri;
        var kind = char.ToUpperInvariant(parts[1][0]) + parts[1][1..];
        return $"{kind} · {parts[2]}";
    }
}
